web/webkit/src/main/scala/net/liftweb/http/LiftSession.scala

```diff
@@ -579,12 +603,18 @@ class LiftSession(private[http] val _contextPath: String, val underlyingId: Stri
    * Updates the internal functions mapping
    */
   def updateFunctionMap(funcs: Map[String, S.AFuncHolder], uniqueId: String, when: Long): Unit = {
-    funcs.foreach {
+    copyFunctions(funcs, nmessageCallback)
+
+    if(LiftRules.putAjaxFnsInContainerSession)
+      updateHttpSessionFns(copyFunctions(funcs, _))
+  }
+
+  private def copyFunctions(from: Map[String, S.AFuncHolder], to:java.util.Map[String, S.AFuncHolder]):Unit =
+    from.foreach {
       case (name, func) =>
-        nmessageCallback.put(name,
+        to.put(name,
           if (func.owner == Full(uniqueId)) func else func.duplicate(uniqueId))
     }
```

Comment:
I want to point out that I originally wanted to put the entire `nmessageCallback` in the session, but unfortunately it contains massive amounts of non-serializable stuff. IIRC [this particular `put`](https://github.com/lift/framework/blob/joe-ajax-fns-in-session/web/webkit/src/main/scala/net/liftweb/http/LiftSession.scala#L995-L997) was a notable culprit.  So there is more that we could theoretically support for failover if we meticulously groom the object tree that is creeping into this map.